core/src/main/java/org/mule/runtime/core/api/exception/DefaultErrorTypeRepository.java

```diff
@@ -88,27 +90,38 @@ public DefaultErrorTypeRepository() {
    */
   @Override
   public ErrorType addErrorType(ComponentIdentifier errorTypeIdentifier, ErrorType parentErrorType) {
-    return addErrorTypeTo(errorTypeIdentifier, parentErrorType, this.errorTypes);
+    checkIfErrorIsAlreadyRegistered(errorTypeIdentifier, ImmutableMap.<ComponentIdentifier, ErrorType>builder()
+        .putAll(errorTypes)
+        .putAll(internalErrorTypes)
+        .build());
+    ErrorType errorType = buildErrorType(errorTypeIdentifier, parentErrorType);
+    errorTypes.put(errorTypeIdentifier, errorType);
+    return errorType;
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public ErrorType addInternalErrorType(ComponentIdentifier errorTypeIdentifier, ErrorType parentErrorType) {
-    return addErrorTypeTo(errorTypeIdentifier, parentErrorType, this.internalErrorTypes);
+    checkIfErrorIsAlreadyRegistered(errorTypeIdentifier, internalErrorTypes);
```

Comment:
Shouldn't we check both maps each time? We won't need collapsing the maps into one then as above.